python/cuspatial/cuspatial/geometry/geocolumn.py

```diff
@@ -0,0 +1,491 @@
+            item_start = multi_end + preceding_line_count
+        else:
+            item_start = preceding_line_count
+        item_length = self._source._lengths[self._index]
+        item_end = item_length + item_start
+        item_source = self._source.lines
+        result = item_source[item_start:item_end]
+        return LineString(
+            np.array(result).reshape(2 * (item_start - item_end), 2)
+        )
+
+
+class gpuMultiLineString(gpuGeometry):
+    def to_shapely(self):
+        item_type = self._source._types[self._index]
+        index = 0
+        for i in range(self._index):
+            if self._source._types[i] == item_type:
+                index = index + 1
+        line_indices = slice(
+            self._source.lines.mlines[index * 2],
+            self._source.lines.mlines[index * 2 + 1],
+        )
+        return MultiLineString(
+            [
+                LineString(
+                    np.array(self._source.lines[i]).reshape(
+                        int(len(self._source.lines[i]) / 2), 2
+                    )
+                )
+                for i in range(line_indices.start, line_indices.stop, 1)
+            ]
+        )
+
+
+class gpuPolygon(gpuGeometry):
+    def to_shapely(self):
+        mp_index = self._index - 1
+        preceding_poly_count = 0
+        preceding_mp_count = 0
+        while mp_index >= 0:
+            if self._source._types[mp_index] == "mpoly":
+                preceding_mp_count = preceding_mp_count + 1
+            elif (
+                self._source._types[mp_index] == "poly"
+                and preceding_mp_count == 0
+            ):
+                preceding_poly_count = preceding_poly_count + 1
+            mp_index = mp_index - 1
+        preceding_multis = preceding_mp_count
+        multi_index = (
+            self._source.polygons.mpolys[preceding_multis * 2 - 1]
+            if preceding_multis > 0
+            else 0
+        )
+        preceding_polys = preceding_poly_count
+        ring_start = self._source.polygons.polys[multi_index + preceding_polys]
+        ring_end = self._source.polygons.polys[
+            multi_index + preceding_polys + 1
+        ]
+        rings = self._source.polygons.rings
+        exterior_slice = slice(rings[ring_start], rings[ring_start + 1])
+        exterior = self._source.polygons.xy[exterior_slice]
+        return Polygon(
+            np.array(exterior).reshape(2 * (ring_start - ring_end), 2),
+            [
+                np.array(self._source.polygons.xy[interior_slice]).reshape(
+                    int((interior_slice.stop - interior_slice.start + 1) / 2),
+                    2,
+                )
+                for interior_slice in [
+                    slice(rings[interior], rings[interior + 1])
+                    for interior in range(ring_end - (ring_start + 1))
+                    + (ring_start + 1)
+                ]
+            ],
+        )
```

Comment:
A comment explaining the math here could be helpful for posterity :) Ditto for `MultiPolygon` below.

Reply:
Done